File: sdks/ios/Sources/RajutechieStreamKit/Transport/RetryPolicy.swift
import Foundation

public struct RetryPolicy {
    public let maxRetries: Int
    public let baseDelay: TimeInterval
    public let maxDelay: TimeInterval
    public let jitter: Bool

    public init(maxRetries: Int = 3, baseDelay: TimeInterval = 1.0, maxDelay: TimeInterval = 30.0, jitter: Bool = true) {
        self.maxRetries = maxRetries
        self.baseDelay = baseDelay
        self.maxDelay = maxDelay
        self.jitter = jitter
    }

    public func execute<T>(_ operation: () async throws -> T) async throws -> T {
        var lastError: Error?

        for attempt in 0...maxRetries {
            do {
                return try await operation()
            } catch {
                lastError = error
                if attempt >= maxRetries { break }

                let exponential = baseDelay * pow(2.0, Double(attempt))
                let capped = min(exponential, maxDelay)
                let delay = jitter ? capped * (0.5 + Double.random(in: 0...0.5)) : capped
                try await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
            }
        }

        throw lastError ?? RajutechieStreamKitError.unknown
    }
}

File: sdks/android/streamkit-sdk/src/main/kotlin/com/rajutechie/streamkit/transport/RetryPolicy.kt
package com.rajutechie.streamkit.transport

import kotlinx.coroutines.delay
import kotlin.math.min
import kotlin.random.Random

class RetryPolicy(
    private val maxRetries: Int = 3,
    private val baseDelay: Long = 1000L,
    private val maxDelay: Long = 30000L,
    private val jitter: Boolean = true
) {
    suspend fun <T> execute(block: suspend () -> T): T {
        var lastException: Exception? = null

        for (attempt in 0..maxRetries) {
            try {
                return block()
            } catch (e: Exception) {
                lastException = e
                if (attempt >= maxRetries) break

                val exponential = baseDelay * (1L shl attempt)
                val capped = min(exponential, maxDelay)
                val delayMs = if (jitter) (capped * (0.5 + Random.nextDouble() * 0.5)).toLong() else capped
                delay(delayMs)
            }
        }

        throw lastException ?: Exception("Retry exhausted")
    }
}
